docs/proposals/20200220-cluster-resource-set.md

```diff
@@ -0,0 +1,202 @@
+Users today must first remember to add these things to every cluster they create, and then actually add them.
+
+Having a mechanism to sync an initial set of default resources after clusters are created makes clusters created with Cluster API functional and ready for workloads from the beginning, without requiring additional user intervention. 
+
+### Goals
+
+Provide a means to specify a set of resources to apply automatically to newly-created and existing Clusters. Resources initially will be applied only once, adding sync functionality is optional
+Support additions to the resource list by applying the new added resources to both new and existing matching clusters
+Support both json and yaml resources
+
+### Non-Goals/Future Work
+
+Replace or compete with the Cluster Addons subproject
+Support deletion of resources. Deleting resources from the list will not result in deletion of those resources from the synced clusters, but deleted resources will not be applied to the new clusters
+Responsible for cleaning up resources applied by ClusterResourceSet controller when ClusterResourceSet resource is deleted
+Responsible for lifecycle management of the installed resources (such as CNI)
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As someone creating multiple clusters, I want all my clusters to have a CNI provider of my choosing installed automatically, so I don’t have to manually repeat the installation for each new cluster.
+
+#### Story 2
+
+As someone creating multiple clusters, I want all my clusters to have a StorageClass installed automatically, so I don't have to manually repeat the installation for each new cluster.
+
+#### Story 3
+
+As someone creating multiple clusters, I want to be able to provide different values for some fields in the resources for different clusters. For example, CNIs podCIDRs  may be required to be distinct for each cluster, hence some templating mechanism for variable substitution in the resources is needed.
+
+### Implementation Details/Notes/Constraints
+
+#### Data model changes
+
+None. We are planning to implement this feature without modifying any of the existing structure to minimize the footprint of ClusterResourceSet Controller. This enhancement will follow Kubernetes’s feature-gate structure and will be under the experimental package with its APIs, and enabled/disabled with a feature gate. 
+
+#### ClusterResourceSet Object Definition
+
+This is the CRD that is used to have a set of components that will be applied to clusters that match the label selector in it.
+
+The resources field is a list of secrets/configmaps in the same namespace. As a cluster label selector, any key-value pair works as long as the same key-value label is assigned to clusters that the addon will be applied to. The reason not to use a predefined label here is to allow matching with multiple ClusterResourceSet objects.
```

Comment:
```suggestion
The resources field is a list of secrets/configmaps in the same namespace. The clusterSelector field is a Kubernetes [label selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements) that matches against labels on clusters.
```

Reply:
Is the label selector in the ClusterResourceSet applied to all clusters in all namespaces, or only to clusters in the same namespace as the ClusterResourceSet?

Reply:
All namespaces. If it is only applied to the clusters in the same namespace as ClusterResourceSet, users may end up creating same resources (secrets/configmaps and ClusterResourceSets) over and over again for clusters in different namespaces.

Reply:
Ok, this needs to be listed explicitly in the proposal. And it also means that if you have permissions to create a ClusterResourceSet but **not** to get secrets or configmaps, you can get access to secrets/configmaps by creating a Cluster + ClusterResourceSet (you'd be able to read the secrets/configmaps in the workload cluster after it's created). We probably need to deal with this.

Reply:
Good point @ncdc. Is there a way to check if the ClusterResourceSet creator has access to resources?

Reply:
Not really

Reply:
@enj is there any way we can safely do impersonation here using a service account?

Reply:
@ncdc not sure if impersonation helps here.  If you can wire a `ValidatingWebhookConfiguration` to this resource on `create`/`update` calls, you could just do two `SubjectAccessReview`s on the user for `list` on `secrets`/`configmaps` cluster-wide.  Thus to use this API, you would need RBAC at the cluster level to mutate it, and full read on all secrets and configs maps in the cluster.

I would like to point out that this resource effectively requires near `cluster-admin` level access because of the ability to read any secret in the cluster.  If you expect it to be commonly used, you are effectively requiring all cluster provisioners to be `cluster-admin`.  Is that okay?

Reply:
> you are effectively requiring all cluster provisioners to be cluster-admin. Is that okay?
In the case of a self-managed management cluster, which is pretty likely, that would be fine, since the controllers are already inherently cluster-admin at that stage.

I wonder, if you should do the same as I've done in the CAPA multi-tenancy proposal, and steal/copy Service API Evolution, and add a cluster-level ClusterResourceSet resource (will need a better name than ClusterClusterResourceSet), adding the `allowedNamespaces` field. This would in turn only read secrets from a specific namespace (cluster-api-system?), overridable at the controller configuration level.

The standard ClusterResourceSet can then be namespace-scoped resource and restricted to only secrets and clusters in the same namespace.

Reply:
A two tiered approach seems reasonable.